### DISCUSSION TITLE
snipe-it auto-loggs out other sessions

### DISCUSSION THREAD

**Discussion by axkibe** [snipe/snipe-it#14126] [Q&A] [answered]
In our installation snipe-it whenever I log in one browser, it will always logout all other browser sessions. This is mildly annoying, as for example using a desktop to make changes and a smartphone as QR-Code reader to identify devices.

I cannot tell if this is by design, by configuration or by accident.

I looked at the source code for a few hours and cannot see anything that should do this on purpose, but with laravel this gets through a few layers of code and I'm a total newbe. The only instance I found on purpose "Auth::logoutOtherDevices" is on passwort change.. which makes a lot of sense, but why does it already auto logout any old session on login?

Is this on purpose (which can be turned off?), or is this misconfiguration on our side perhaps?

**Comment by snipe** [comment]
That's definitely not a thing we're doing. There is an env var (I don't know who or how your stuff was set up): https://snipe-it.readme.io/docs/configuration#optional-session-settings `EXPIRE_ON_CLOSE` and another one `SESSION_LIFETIME` you might want to look at, but's a config issue or possible a network issue (if you have clever network people trying to prevent shadow IT)

**Reply by gettalong** [reply]
@snipe I'm a colleague of @axkibe. I have checked the `.env` file and `EXPIRE_ON_CLOSE` is false while the `SESSION_LIFETIME` is 12000. So this should not be a problem.

Additionally, the problem can be reproduced on the same machine with the same browser, logging in once in a standard tab and once in private tab. As soon as I log in one of the tabs while the other is already logged in, the other one gets logged out. So this should not be a network issue since the server is accessed with the same IP (verified via Apache log files).

With the `APP_DEBUG` setting enabled, I ran the above tests, too, but didn't find anything in the Laravel debugbar or in storage/logs/laravel.log that would indicate why this happens.

Is there something we can do to find the culprit? Like putting debug statements into certain parts of the code to see where the redirect to the login page happens?

I tried to find the right place, alas, I'm not really familiar with Laravel.

**Reply by gettalong** [reply]
Just writing again since this is still not solved. Help in trying to get to the root of the problem would be greatly appreciated. As written before a pointer to the location in the code where it is checked whether someone is already logged in would already help.

**Reply by snipe** [reply]
- https://github.com/snipe/snipe-it/blob/master/app/Http/Controllers/Auth/LoginController.php handles the login
- https://github.com/snipe/snipe-it/blob/master/app/Http/Kernel.php - points to the library middleware that handles authorization and authentication

**Reply by gettalong** [reply]
Thanks! That helped a bit. I was able to confirm that including `\Illuminate\Auth\Middleware\Authenticate::class` the `request->user()` object is available and correct. So something after that must log out the user and perform the redirect to login. What is run after that last middleware?

**Reply by snipe** [reply]
Just the normal application stack. The only place we invoke a logout is here:

https://github.com/snipe/snipe-it/blob/931ab103c1c3aac360fd362a64b0aa067a4ce4e2/app/Http/Controllers/Auth/LoginController.php#L452-L499

And on a restore (where everyone gets logged out), and in the `CheckUserIsActivated middleware`.

**Reply by gettalong** [reply, marked as answer]
Thanks again for the pointers! I found the culprit.

Using `php artisan route:list -vvv --path=admin` I found that after `\Illuminate\Auth\Middleware\Authenticate` the `Illuminate\Session\Middleware\AuthenticateSession` middleware is run before the application stack.

And inside the `handle()` method the following part logs the user out:

https://github.com/laravel/framework/blob/11.x/src/Illuminate/Session/Middleware/AuthenticateSession.php#L63-L65

I'm not sure why the password hash changes, though, since the password didn't change. **However, after disabling "LDAP Password Sync" it now works and I can login multiple times concurrently!**

**Reply by snipe** [reply]
Great sleuthing! We wouldn't know that the password didn't change though, since if you bcrypt the same text 100, the hash will change each time. We honestly may deprecate LDAP password syncing anyway. Most LDAP systems are pretty stable these days, and since most are in the cloud for many of our users, if they cannot contact the LDAP system, they have bigger problems anyway.
